include/pybind11/pybind11.h

```diff
@@ -339,60 +339,71 @@ class cpp_function : public function {
-
-                        if (value)
-                            kwargs_consumed++;
-                        else if (it2.value)
-                            value = it2.value;
-
-                        if (value) {
-                            PyTuple_SET_ITEM(args_.ptr(), index, value.inc_ref().ptr());
-                        } else {
-                            kwargs_consumed = (size_t) -1; /* definite failure */
-                            break;
+                        int arg_ctr = 0;
+                        for (auto const &it2 : it->args) {
+                            int index = arg_ctr++;
+                            if (PyTuple_GET_ITEM(args_.ptr(), index))
+                                continue;
+
+                            handle value;
+                            if (kwargs)
+                                value = PyDict_GetItemString(kwargs, it2.name);
+
+                            if (value)
+                                kwargs_consumed++;
+                            else if (it2.value)
+                                value = it2.value;
+
+                            if (value) {
+                                PyTuple_SET_ITEM(args_.ptr(), index, value.inc_ref().ptr());
+                            } else {
+                                kwargs_consumed = (size_t) -1; /* definite failure */
+                                break;
+                            }
                         }
                     }
-                }
 
-                try {
-                    if ((kwargs_consumed == nkwargs || it->has_kwargs) &&
-                        (nargs_ == it->nargs || it->has_args))
-                        result = it->impl(it, args_, kwargs, parent);
-                } catch (cast_error &) {
-                    result = PYBIND11_TRY_NEXT_OVERLOAD;
-                }
+                    try {
+                        if ((kwargs_consumed == nkwargs || it->has_kwargs) &&
+                            (nargs_ == it->nargs || it->has_args))
+                            result = it->impl(it, args_, kwargs, parent);
+                    } catch (cast_error &) {
+                        result = PYBIND11_TRY_NEXT_OVERLOAD;
+                    }
 
-                if (result.ptr() != PYBIND11_TRY_NEXT_OVERLOAD)
-                    break;
+                    if (result.ptr() != PYBIND11_TRY_NEXT_OVERLOAD)
+                        break;
+                }
+            } catch (...) {
+                // Give any registered exception handler a chance first (in reverse order of registration)
+                auto &registered_exception_translators = pybind11::detail::get_internals().registered_exception_translators;
+                for (auto& translator : registered_exception_translators) {
+                    if (translator(std::current_exception())) return nullptr;
+                }
+                // If we get here, delegate to standard exception handlers
+                throw; 
```

Comment:
I think you might want to add a copy of the catch block for `error_already_set` into this `try` and give it higher priority than the custom exception handlers; since it already contains a translated Python exception, I don't think we want to give the custom handlers the opportunity to do something with it.